frontend/src/views/TE/Upload.vue

```diff
@@ -190,6 +209,62 @@ export default class Upload extends Vue {
     }
   }
 
+  getFileHash(file: File) {
+    return new Promise(resolve => {
+      const fileReader = new FileReader();
+      fileReader.onload = function(event: any) {
+        const spark = new Spark.ArrayBuffer();
+        spark.append(event.target.result);
+        resolve(btoa(spark.end(true)));
+      };
+      fileReader.readAsArrayBuffer(file);
+    });
+  }
+
+  async validateFileHash(file: File, task: any) {
+    const fileHash = await this.getFileHash(file);
+    if (!task.versions) return;
+    for (const version of Object.values<any>(task.versions)) {
+      const ref = this.uploadsBucket.ref().child(`/${version.uploadPath}`);
+      const metadata = await ref.getMetadata().catch(e => "error");
+      if (metadata === "error") continue;
+      if (fileHash === metadata.md5Hash) {
+        throw new Error("You had uploaded the same file earlier.");
+      }
+    }
+  }
+
+  async validateUpload(file: File) {
+    validateFlacFile(file);
+    const taskId = getTaskId(file.name);
+    const task = (
+      await firebase
+        .database()
+        .ref(`/TE/tasks/${taskId}`)
+        .once("value")
+    ).val();
+    if (this.currentUser.email !== _.get(task, "assignee.emailAddress")) {
+      throw new Error(`The task ${taskId} is not assigned to you.`);
+    }
```

Comment:
According to the database rules, the user will be able to read the task only if she is an assignee, so this check is redundant. Instead, you should validate that you have not got access violation.

Reply:
What would be an access violation? Someone who is not an editor or coordinator or is not assigned to the task?

Reply:
I'm not sure I understand the question.
Currently according to the database rules, only assignee, coordinator and checker can access a task.

Reply:
Sorry, my question was poorly phrased. I understand now though. An editor can also upload files should they be allowed access too?

Reply:
Assignee = editor, who is actually assigned to the current task.